src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -219,50 +218,47 @@ void RemoveProject(Project project)
     /// </remarks>
     private protected async virtual ValueTask ProcessWorkAsync(ImmutableArray<Work> work, CancellationToken cancellationToken)
     {
-        // Capture as locals here. Cancellation of the work queue still need to propogate. The cancellation
-        // token itself represents the work queue halting, but this will help avoid any assumptions about nullability of locals
-        // through the use in this function.
-        var stream = _stream;
-        var solution = _workspace?.CurrentSolution;
-
-        cancellationToken.ThrowIfCancellationRequested();
-
-        // Early bail check for if we are disposed or somewhere in the middle of disposal 
-        if (_disposed || stream is null || solution is null)
+        if (cancellationToken.IsCancellationRequested)
```

Comment:
Call me paranoid, but I still have flashbacks to the last time a disposed flag was changed to check a cancellation token, and that didn't go well.

If I'm reading this right, the assumption is that `stream` will never be null because it's only set to null in `DIspose`, and we're checking cancellation here. BUT unless I missed it, the `AsyncBatchingWorkQueue` doesn't create a linked cancellation token for an individual batch, and the whole queue. I _think_ it might be possible for us to be disposed, and not necessarily have _this_ token be cancelled.

If I'm wrong, then great, but perhaps a comment explaining why it can't happen would be good. Alternatively, just checking both tokens here might make sense.

Reply:
> BUT unless I missed it, the AsyncBatchingWorkQueue doesn't create a linked cancellation token for an individual batch, and the whole queue.

The `AsyncBatchingWorkQueue` does create a linked cancellation token so that a batch will be cancelled when the main cancellation token is cancelled.  It uses a `CancellationSeries` to do exactly that. So, if `_disposeTokenSource.Cancel()` is called, the cancellation token passed to the batch will also be in a cancelled state.

Here's the relevant code in `AsyncBatchingWorkQueue`:

https://github.com/dotnet/razor/blob/07e138222c3b32bcba863b346fda8321fa62c82f/src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/AsyncBatchingWorkQueue%602.cs#L105-L138

`CancellationService.CreateNext()` does the work of creating a linked token source.